File: PHP/FoodTruck/controller/Controller.php
<?php
require_once __DIR__.'/InputValidator.php';
require_once __DIR__.'/../persistence/PersistenceFoodTruck.php';
require_once __DIR__.'/../model/Equipment.php';
require_once __DIR__.'/../model/Supply.php';
require_once __DIR__.'/../model/MenuItem.php';
require_once __DIR__.'/../model/Order.php';
require_once __DIR__.'/../model/TimeBlock.php';
require_once __DIR__.'/../model/Staff.php';


require_once __DIR__.'/../model/FTMS.php';

class Controller{

	public function __construct(){

	}
//TODO: length evaluation (string less than 50 etc)
	public function createEquipment($equipment_name, $equipment_quantity){
		//1. Validate input
		$error = "";
		$name = InputValidator::validate_input($equipment_name);
		$quantity = InputValidator::validate_input($equipment_quantity);
		$isInt = preg_match('/^[0-9]+$/',$quantity);
		if ($name!=null && strlen($name)!=0 && strlen($name) <=50 && $quantity!=null && strlen($quantity)!=0 && strlen($quantity)<=9 && $isInt){
			//2. load all of the data
			$pm = new PersistenceFoodTruck();
			$rm = $pm->loadDataFromStore();
		
			//3. Add the new Equipment
			$equipment = new Equipment($name,intval($quantity));
			$rm->addEquipment($equipment);
		
			//4. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else {
			if ($name == null || strlen($name)==0) {
				$error .= "@1Equipment name cannot be empty! ";
			}
			else if (strlen($name) > 50){
				$error .="@1Equipment name cannot be longer than 50 characters! ";
			}
			if ($quantity == null || strlen($quantity) ==0) {
				$error .= "@2Equipment quantity cannot be empty! ";
			}
			else if (!$isInt) {
				$error .= "@2Equipment quantity must be a positive integer! ";
			}
			else if(strlen($quantity) > 9){
				$error .= "@2Equipment quantity cannot have more than 9 digits! ";
			}
			throw new Exception(trim($error));
		} 
	}
	
	public function createSupply($supply_name, $supply_quantity){
		//1. Validate input
		$error = "";
		$name = InputValidator::validate_input($supply_name);
		$quantity = InputValidator::validate_input($supply_quantity);
		$isInt = preg_match('/^[0-9]+$/',$quantity);
		if ($name!=null && strlen($name)!=0 && strlen($name) <=50 && $quantity!=null && strlen($quantity)!=0 && strlen($quantity) <=9 && $isInt){
			//2. load all of the data
			$pm = new PersistenceFoodTruck();
			$rm = $pm->loadDataFromStore();
	
			//3. Add the new Supply
			$supply = new Supply($name,intval($quantity));
			$rm->addSupply($supply);
	
			//4. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else {
			if ($name == null || strlen($name)==0) {
				$error .= "@1Supply name cannot be empty! ";
			}
			else if (strlen($name) > 50){
				$error .="@1Supply name cannot be longer than 50 characters! ";
			}
			if ($quantity == null || strlen($quantity) ==0) {
				$error .= "@2Supply quantity cannot be empty! ";
			}
			else if (!$isInt) {
				$error .= "@2Supply quantity must be a positive integer! ";
			}
			else if (strlen($quantity) > 9){
				$error .="@2Supply quantity cannot have more than 9 digits! ";
			}
			throw new Exception(trim($error));
		}
	}

	public function createStaff($staff_name, $staff_role){
		//1. Validate input
		$error = "";
		$name = InputValidator::validate_input($staff_name);
		$role = InputValidator::validate_input($staff_role);
		
		//2. load all of the data
		$pm = new PersistenceFoodTruck();
		$rm = $pm->loadDataFromStore();
		
		if ($name!=null && strlen($name)!=0 && strlen($name) <=50 && $role!=null && strlen($role)!=0 && strlen($role) <=50){
			
	
			//3. Add the new Staff
			$staff = new Staff($name,$role);
			$rm->addStaff($staff);
	
			//4. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else {
			if ($name == null || strlen($name)==0) {
				$error .= "@1Staff name cannot be empty! ";
			}
			else if (strlen($name) > 50){
				$error .="@1Staff name cannot be longer than 50 characters! ";
			}
			if ($role == null || strlen($role) ==0) {
				$error .= "@2Staff role cannot be empty! ";
			}
			else if (strlen($name) > 50){
				$error .="@2Staff role cannot be longer than 50 characters! ";
			}
			
			throw new Exception(trim($error));
		}
	}
	
	public function createTimeBlock($block_starttime, $block_endtime, $block_dayofweek,$block_staff){
		//1. Validate input
		$error = "";
		$day = InputValidator::validate_input($block_dayofweek);
		$s = preg_match("/(2[0-3]|[01][0-9]):([0-5][0-9])/", $block_starttime);
		$e = preg_match("/(2[0-3]|[01][0-9]):([0-5][0-9])/", $block_endtime);
		
		//2. load all of the data
		$pm = new PersistenceFoodTruck();
		$rm = $pm->loadDataFromStore();
		
		//3. Find staff
		$myKey = -1;
		foreach ($rm->getStaffs() as $key => $staff){
			if (strcmp($staff->getName(), $block_staff)==0){
				$myKey = $key;
				break;
			}
		}
		
		if ($day != null && strlen($day)!= 0 && $s && $e && $block_starttime<=$block_endtime && $myKey != -1){
			//4. Add the new time block
			$timeblock = new TimeBlock($block_starttime, $block_endtime, $day);
			$rm->addTimeBlock($timeblock);
			
			//5. Add time block to staff
			$rm->getStaff_index($myKey)->addTimeBlock($timeblock);	
			
			//6. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else{
			if($day == null || strlen($day) ==0){
				$error .= "@1Time block day of the week cannot be empty! ";
			}
			if (!$s){
				$error .= "@2Time block start time must be specified correctly (HH:MM)! ";
			}
			if (!$e) {
				$error .= "@3Time block end time must be specified correctly (HH:MM)! ";
			}
			if ($block_endtime < $block_starttime){
				$error .= "@3Time block end time cannot be before event start time! ";
			}
			if($myKey == -1){
				$error .= "@4Staff not found! ";
			}
			throw new Exception(trim($error));
		}
	}
	
	public function createMenuItem($item_name, $item_supplies){	
		//1. load all of the data
		$pm = new PersistenceFoodTruck();
		$rm = $pm -> loadDataFromStore();
		
		//2. Validate input
		$error = "";
		$name = InputValidator::validate_input($item_name);
		
		//3. Find supplies
		$is_supplies = False;
		foreach ($rm->getSupplies() as $supply){
			foreach ($item_supplies as $supplyName){
				if (strcmp($supply->getName(), $supplyName)==0){
					$is_supplies = True;
					break;
				}
			}
		}
		
		if ($name!=null && strlen($name)!=0 && strlen($name)<=50 && $is_supplies){
			//4. Add the new Menu Item
			$menuItem = new MenuItem($name);
			
			//5. Find supplies again and link to Menu Item
			foreach ($rm->getSupplies() as $supply){
				foreach ($item_supplies as $supplyName){
					if (strcmp($supply->getName(), $supplyName)==0){
						$menuItem->addSupply($supply);
					}
				}
			}
			$rm->addMenuItem($menuItem);
	
			//6. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else {
			if ($name == null || strlen($name)==0) {
				$error .= "@1Menu item name cannot be empty! ";
			}
			else if (strlen($name) > 50){
				$error .="@1Menu item name cannot be longer than 50 characters! ";
			}
			if (!$is_supplies) {
				$error .= "@2Menu item ingredient not found! ";
			}
				
			throw new Exception(trim($error));
		}
	}
	
	public function createOrder($order_menuItem){
		//1. load all of the data
		$pm = new PersistenceFoodTruck();
		$rm = $pm -> loadDataFromStore();
	
		//2. Validate input
		$error = "";
		
		//3. Find menu items
		$is_menuitem = False;
		foreach ($rm->getMenuItems() as $menuitem){
			foreach ($order_menuItem as $itemName){
				if (strcmp($menuitem->getName(), $itemName)==0){
					$is_menuitem = True;
					break;
				}
			}
		}
	
		if ($is_menuitem){
			//4. Add the new Order
			$order = new Order();
				
			//5. Find menuitems again and link to order
			foreach ($rm->getMenuItems() as $menuitem){
				foreach ($order_menuItem as $itemName){
					if (strcmp($menuitem->getName(), $itemName)==0){
						$order->addMenuItem($menuitem);
					}
				}
			}
			$rm->addOrder($order);
	
			//6. Write all of the data
			$pm->writeDataToStore($rm);
		}
		else {
			if (!$is_menuitem) {
				$error .= "Order Menu item not found!";
			}
	
			throw new Exception($error);
		}
	}
}

File: Java/FoodTruck/src/controller/Controller.java
package controller;

import java.sql.Time;
import java.util.ArrayList;

import controller.InvalidInputException;
import model.Equipment;
import model.FTMS;
import model.MenuItem;
import model.Order;
import model.Staff;
import model.Supply;
import model.TimeBlock;
import persistence.PersistenceXStream;

public class Controller {

	public Controller(){

	}

	public void createEquipment(String name, String description, String nonValidatedQuantity) throws InvalidInputException{
		String error = "";
		int numZero = 0;
		boolean valid = true;
		String validatedQuantity = "";
		int quantity = 0;
		if(name == null || name.trim().length() == 0)
			error += "Equipment name cannot be empty! ";
		else 	if(name.length() > 50)
			error += "Equipment name cannot be too long! ";
		if(description == null)	// if user does not input anything for description, use default empty string
			description = "";
		else if(description.length() > 100)
			error += "Equipment description cannot be too long! ";
		if(nonValidatedQuantity == null || nonValidatedQuantity.trim().length() == 0)
			error += "Equipment quantity cannot be empty! ";
		else{
			for(int x = 0; x < nonValidatedQuantity.length(); x++){	// check if quantity is in the form of an integer
				if(nonValidatedQuantity.charAt(x) < '0' || nonValidatedQuantity.charAt(x) > '9'){
					error += "Equipment quantity must be an integer!";
					valid = false;
					break;
				}
			}
			if(valid){
				// cut all leading useless zeros
				while(numZero < nonValidatedQuantity.length()){
					if(nonValidatedQuantity.charAt(numZero) == '0')
						numZero++;
					else
						break;
				}
				if(numZero == nonValidatedQuantity.length())	// if input is all zeros then set it to zero
					validatedQuantity = "0";

				for(int i =0; i < nonValidatedQuantity.length() - numZero; i++){
					validatedQuantity += nonValidatedQuantity.charAt(i + numZero);
				}
				if(validatedQuantity.length() < 9){	// prevent integer out of range error
					quantity = Integer.parseInt(validatedQuantity);
				}else{
					error += "Equipment quantity cannot be too large!";
					valid = false;
				}
			}
		}
		error = error.trim();
		if(error.length() > 0)
			throw new InvalidInputException(error);
		Equipment equipment = new Equipment(name, description, quantity);
		FTMS master = FTMS.getInstance();
		master.addEquipment(equipment);
		PersistenceXStream.saveToXMLwithXStream(master);
	}

	public void createSupply(String name, String description, String nonValidatedQuantity) throws InvalidInputException{
		String error = "";
		int numZero = 0;
		boolean valid = true;
		String validatedQuantity = "";
		int quantity = 0;
		if(name == null || name.trim().length() == 0)
			error += "Supply name cannot be empty! ";
		else 	if(name.length() > 50)
			error += "Supply name cannot be too long! ";
		if(description == null)	// if user does not input anything for description, use default empty string
			description = "";
		else if(description.length() > 100)
			error += "Supply description cannot be too long! ";
		if(nonValidatedQuantity == null || nonValidatedQuantity.trim().length() == 0)
			error += "Supply quantity cannot be empty! ";
		else{
			for(int x = 0; x < nonValidatedQuantity.length(); x++){	// check if quantity is in the form of an integer
				if(nonValidatedQuantity.charAt(x) < '0' || nonValidatedQuantity.charAt(x) > '9'){
					error += "Supply quantity must be an integer!";
					valid = false;
					break;
				}
			}
			if(valid){
				// cut all leading useless zeros
				while(numZero < nonValidatedQuantity.length()){
					if(nonValidatedQuantity.charAt(numZero) == '0')
						numZero++;
					else
						break;
				}
				if(numZero == nonValidatedQuantity.length())	// if input is all zeros then set it to zero
					validatedQuantity = "0";

				for(int i =0; i < nonValidatedQuantity.length() - numZero; i++){
					validatedQuantity += nonValidatedQuantity.charAt(i + numZero);
				}
				if(validatedQuantity.length() < 9){	// prevent integer out of range error
					quantity = Integer.parseInt(validatedQuantity);
				}else{
					error += "Supply quantity cannot be too large!";
					valid = false;
				}
			}
		}
		error = error.trim();
		if(error.length() > 0)
			throw new InvalidInputException(error);
		Supply supply = new Supply(name, description, quantity);
		FTMS master = FTMS.getInstance();
		master.addSupply(supply);
		PersistenceXStream.saveToXMLwithXStream(master);
	}

	public void createStaff(String name, String role) throws InvalidInputException{
		String error = "";
		if(name == null || name.trim().length() == 0)
			error += "Staff name cannot be empty! ";
		else 	if(name.length() > 50)
			error += "Staff name cannot be too long! ";
		if(role == null || role.trim().length() == 0)
			error += "Staff role cannot be empty!";
		else if(role.length() > 50)
			error += "Staff role cannot be too long! ";
		error = error.trim();
		if(error.length() > 0)
			throw new InvalidInputException(error);
		Staff staff = new Staff(name, role);
		FTMS master = FTMS.getInstance();
		master.addStaff(staff);
		PersistenceXStream.saveToXMLwithXStream(master);
	}

	public void createMenuItem(String name, ArrayList<Supply> ingredients) throws InvalidInputException{
		String error = "";
		if(name == null || name.trim().length() == 0)
			error += "Menu item name cannot be empty! ";
		else if(name.length() > 50)
			error += "Menu item name cannot be too long! ";
		if(ingredients == null || ingredients.isEmpty())
			error += "Menu item ingredients cannot be empty!";
		else if(ingredients.size() > 50)
			error += "Menu item ingredients cannot exceed 50!";
		error = error.trim();
		if(error.length() > 0)
			throw new InvalidInputException(error);
		MenuItem menuItem = new MenuItem(name);
		for(int x = 0; x < ingredients.size(); x++)
			menuItem.addSupply(ingredients.get(x));
		FTMS master = FTMS.getInstance();
		master.addMenuItem(menuItem);
		PersistenceXStream.saveToXMLwithXStream(master);
	}

	// for dayOfWeek use 1 for Monday, ... , 7 for Sunday
	public void createTimeBlock(Time startTime, Time endTime, int dayOfWeek, Staff staff) throws InvalidInputException{
		FTMS master = FTMS.getInstance();
		String error = "";
		if (startTime == null)
			error += "Time block start time cannot be empty! ";
		if (endTime == null)
			error += "Time block end time cannot be empty! ";
		if (dayOfWeek < 0 || dayOfWeek > 6)
			error += "Time block day of the week must be between Monday and Sunday! ";
		if (startTime != null && endTime != null && endTime.getTime() < startTime.getTime())
			error += "Time block end time cannot be before start time! ";
		if (staff == null)
			error += "Time block must be assigned to a staff! ";
		if (dayOfWeek >= 0 && dayOfWeek <= 6 && staff != null) {
			for (int x = 0; x < master.getStaffs().size(); x++) {
				if (master.getStaff(x).getName() == staff.getName())
					for (int i = 0; i < master.getStaff(x).getTimeBlocks().size(); i++) {
						if (dayOfWeek == master.getStaff(x).getTimeBlock(i).getDayOfWeek() && startTime.toString().equals(master.getStaff(x).getTimeBlock(i).getStartTime().toString())) {
							error += "This time block has already been added! ";
							break;
						}
					}
			}
		}
		error = error.trim();
		if(error.length() > 0)
			throw new InvalidInputException(error);
		TimeBlock timeBlock = new TimeBlock(startTime, endTime, dayOfWeek);
		for(int x = 0; x < master.getStaffs().size(); x++){
			if(master.getStaff(x).getName() == staff.getName())
				master.getStaff(x).addTimeBlock(timeBlock);
		}
		PersistenceXStream.saveToXMLwithXStream(master);
	}

	public void createOrder(ArrayList<MenuItem> menuItems) throws InvalidInputException{
		String error = "";
		if(menuItems.isEmpty())
			error += "You must select at least one menu item! ";
		FTMS master = FTMS.getInstance();
		Order order = new Order();
		for(int x = 0; x < menuItems.size(); x++)
			order.addMenuItem(menuItems.get(x));
		int flag = 0;
		for(int x  = 0; x < order.getMenuItems().size(); x++){
			for(int y = 0; y < order.getMenuItem(x).getSupplies().size(); y++){
				flag = 0;
				for(int i = 0; i < master.getSupplies().size(); i++){
					if(master.getSupply(i).getName() == order.getMenuItem(x).getSupply(y).getName() && master.getSupply(i).getQuantity() > 0){
						master.getSupply(i).setQuantity(master.getSupply(i).getQuantity() - 1);
						flag  = 1;
						break;
					}
				}
				if(flag == 0){
					String txt = String.format("%s is unavailable at the moment!", order.getMenuItem(x).getName());
					error += txt;
					error = error.trim();
					if(error.length() > 0)
						throw new InvalidInputException(error);
				}
			}
		}
		for(int x = 0; x < order.getMenuItems().size(); x++){
			for(int y = 0; y < master.getMenuItems().size(); y++){
				if(master.getMenuItem(y).getName() == order.getMenuItem(x).getName())
					master.getMenuItem(y).setPopularity(master.getMenuItem(y).getPopularity() + 1);
			}
		}

		master.addOrder(order);

		PersistenceXStream.saveToXMLwithXStream(master);
	}
}
